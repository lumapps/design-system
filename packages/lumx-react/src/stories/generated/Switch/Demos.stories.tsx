/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/switch/Demos' };

export { App as Default } from './default';
