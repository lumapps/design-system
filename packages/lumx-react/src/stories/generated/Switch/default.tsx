/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/switch/react/default.tsx