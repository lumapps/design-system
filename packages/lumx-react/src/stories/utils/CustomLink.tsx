/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
