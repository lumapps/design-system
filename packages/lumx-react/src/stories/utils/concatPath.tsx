export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
