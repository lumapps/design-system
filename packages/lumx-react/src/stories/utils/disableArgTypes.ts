export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
