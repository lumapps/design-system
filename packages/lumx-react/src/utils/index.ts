/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
