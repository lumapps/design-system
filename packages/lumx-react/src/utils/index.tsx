export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
