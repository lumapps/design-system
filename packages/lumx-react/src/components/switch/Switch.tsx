import React, { Children, forwardRef, SyntheticEvent, useMemo } from 'react';

import classNames from 'classnames';
import { uid } from 'uid';

import isEmpty from 'lodash/isEmpty';

import { Alignment, InputHelper, InputLabel, Theme } from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface SwitchProps extends GenericProps {
    /** Helper text. */
    helper?: string;
    /** Whether it is checked or not. */
    isChecked?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Native input name property. */
    name?: string;
    /** Position of the switch relative to the label. */
    position?: Extract<Alignment, 'right' | 'left'>;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Native input value property. */
    value?: string;
    /** On change callback. */
    onChange?(isChecked: boolean, value?: string, name?: string, event?: SyntheticEvent): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Switch';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<SwitchProps> = {
    position: Alignment.left,
    theme: Theme.light,
};

/**
 * Switch component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Switch: Comp<SwitchProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        checked,
        children,
        className,
        disabled,
        helper,
        id,
        isChecked = checked,
        isDisabled = disabled,
        name,
        onChange,
        position,
        theme,
        value,
        ...forwardedProps
    } = props;
    const switchId = useMemo(() => id || `switch-${uid()}`, [id]);
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (onChange) {
            onChange(!isChecked, value, name, event);
        }
    };

    return (
        <div
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                    isChecked,
                    isDisabled,
                    position,
                    theme,
                    isUnchecked: !isChecked,
                }),
            )}
            aria-disabled={isDisabled}
        >
            <div className={`${CLASSNAME}__input-wrapper`}>
                <input
                    type="checkbox"
                    role="switch"
                    id={switchId}
                    className={`${CLASSNAME}__input-native`}
                    name={name}
                    value={value}
                    disabled={isDisabled}
                    checked={isChecked}
                    aria-checked={Boolean(isChecked)}
                    onChange={handleChange}
                />

                <div className={`${CLASSNAME}__input-placeholder`}>
                    <div className={`${CLASSNAME}__input-background`} />
                    <div className={`${CLASSNAME}__input-indicator`} />
                </div>
            </div>

            {Children.count(children) > 0 && (
                <div className={`${CLASSNAME}__content`}>
                    <InputLabel htmlFor={switchId} theme={theme} className={`${CLASSNAME}__label`}>
                        {children}
                    </InputLabel>
                    {!isEmpty(helper) && (
                        <InputHelper theme={theme} className={`${CLASSNAME}__helper`}>
                            {helper}
                        </InputHelper>
                    )}
                </div>
            )}
        </div>
    );
});
Switch.displayName = COMPONENT_NAME;
Switch.className = CLASSNAME;
Switch.defaultProps = DEFAULT_PROPS;
