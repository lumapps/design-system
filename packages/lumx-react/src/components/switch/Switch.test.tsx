import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { build, oneOf } from 'test-data-bot';

import without from 'lodash/without';

import { Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { Theme, Alignment } from '@lumx/react';
import { Switch, SwitchProps } from './Switch';

const DEFAULT_PROPS = Switch.defaultProps as any;
const CLASSNAME = Switch.className as string;

type SetupProps = Partial<SwitchProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: SwitchProps = {
        ...propsOverride,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<Switch {...props} />);

    return {
        root: wrapper.find('div').first(),

        input: wrapper.find('input'),
        inputWrapper: wrapper.find(`.${CLASSNAME}__input-wrapper`),

        content: wrapper.find(`.${CLASSNAME}__content`),
        helper: wrapper.find(`.${CLASSNAME}__helper`),
        label: wrapper.find(`.${CLASSNAME}__label`),

        props,
        wrapper,
    };
};

jest.mock('uid', () => ({ uid: () => 'uid' }));

describe(`<${Switch.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly without any label', () => {
            const { root, inputWrapper, input, content, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(root).toExist();
            expect(root).toHaveClassName(CLASSNAME);

            expect(inputWrapper).toExist();
            expect(input).toExist();

            expect(content).not.toExist();
        });

        it('should render correctly with only a `label`', () => {
            const props: SetupProps = { children: 'Label' };
            const { root, inputWrapper, input, content, helper, label, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(root).toExist();
            expect(root).toHaveClassName(CLASSNAME);

            expect(inputWrapper).toExist();
            expect(input).toExist();

            expect(content).toExist();
            expect(label).toExist();
            expect(helper).not.toExist();
        });

        it('should render correctly with a `label` and a `helper`', () => {
            const props: SetupProps = { children: 'Label', helper: 'Helper' };
            const { root, inputWrapper, input, content, helper, label, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(root).toExist();
            expect(root).toHaveClassName(CLASSNAME);

            expect(inputWrapper).toExist();
            expect(input).toExist();

            expect(content).toExist();
            expect(label).toExist();
            expect(helper).toExist();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default props', () => {
            const { root } = setup();

            Object.keys(DEFAULT_PROPS).forEach((prop: string) => {
                let defaultProp: any = DEFAULT_PROPS[prop];

                if (prop === 'checked') {
                    // eslint-disable-next-line no-param-reassign
                    prop = 'unchecked';
                    defaultProp = true;
                }

                expect(root).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: prop, value: defaultProp }));
            });
        });

        it('should use the given props', () => {
            const modifiedPropsBuilder: () => SetupProps = build('props').fields({
                isChecked: true,
                position: oneOf(...without(Object.values(Alignment), DEFAULT_PROPS.position)),
                theme: oneOf(...without(Object.values(Theme), DEFAULT_PROPS.theme)),
            });

            const modifiedProps: SetupProps = modifiedPropsBuilder();

            const { root } = setup({ ...modifiedProps });

            Object.keys(modifiedProps).forEach((prop: string) => {
                if (prop === 'checked') {
                    if (modifiedProps[prop]) {
                        expect(root).toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                        );
                    } else {
                        expect(root).toHaveClassName(
                            getBasicClass({ prefix: CLASSNAME, type: 'unchecked', value: true }),
                        );
                    }
                } else {
                    expect(root).toHaveClassName(
                        getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                    );
                }
            });
        });
    });

    // 3. Test events.
    describe('Events', () => {
        const onChange: jest.Mock = jest.fn();

        beforeEach(() => {
            onChange.mockClear();
        });

        it('should trigger `onChange` when toggled', () => {
            const { input } = setup({ onChange }, false);

            input.simulate('change');
            expect(onChange).toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        it('should not display the `helper` if no `label` is given', () => {
            const props: SetupProps = { helper: 'Helper' };
            const { content, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(content).not.toExist();
        });
    });

    // 5. Test state.
    // N/A

    // Common tests suite.
    commonTestsSuite(setup, { className: 'root', prop: 'root' }, { className: CLASSNAME });
});
