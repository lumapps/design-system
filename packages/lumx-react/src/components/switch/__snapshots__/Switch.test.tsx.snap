// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch> Conditions should not display the \`helper\` if no \`label\` is given 1`] = `
<div
  className="lumx-switch lumx-switch--position-left lumx-switch--theme-light lumx-switch--is-unchecked"
>
  <div
    className="lumx-switch__input-wrapper"
  >
    <input
      aria-checked={false}
      className="lumx-switch__input-native"
      id="switch-uid"
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="lumx-switch__input-placeholder"
    >
      <div
        className="lumx-switch__input-background"
      />
      <div
        className="lumx-switch__input-indicator"
      />
    </div>
  </div>
</div>
`;

exports[`<Switch> Snapshots and structure should render correctly with a \`label\` and a \`helper\` 1`] = `
<div
  className="lumx-switch lumx-switch--position-left lumx-switch--theme-light lumx-switch--is-unchecked"
>
  <div
    className="lumx-switch__input-wrapper"
  >
    <input
      aria-checked={false}
      className="lumx-switch__input-native"
      id="switch-uid"
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="lumx-switch__input-placeholder"
    >
      <div
        className="lumx-switch__input-background"
      />
      <div
        className="lumx-switch__input-indicator"
      />
    </div>
  </div>
  <div
    className="lumx-switch__content"
  >
    <InputLabel
      className="lumx-switch__label"
      htmlFor="switch-uid"
      theme="light"
    >
      Label
    </InputLabel>
    <InputHelper
      className="lumx-switch__helper"
      kind="info"
      theme="light"
    >
      Helper
    </InputHelper>
  </div>
</div>
`;

exports[`<Switch> Snapshots and structure should render correctly with only a \`label\` 1`] = `
<div
  className="lumx-switch lumx-switch--position-left lumx-switch--theme-light lumx-switch--is-unchecked"
>
  <div
    className="lumx-switch__input-wrapper"
  >
    <input
      aria-checked={false}
      className="lumx-switch__input-native"
      id="switch-uid"
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="lumx-switch__input-placeholder"
    >
      <div
        className="lumx-switch__input-background"
      />
      <div
        className="lumx-switch__input-indicator"
      />
    </div>
  </div>
  <div
    className="lumx-switch__content"
  >
    <InputLabel
      className="lumx-switch__label"
      htmlFor="switch-uid"
      theme="light"
    >
      Label
    </InputLabel>
  </div>
</div>
`;

exports[`<Switch> Snapshots and structure should render correctly without any label 1`] = `
<div
  className="lumx-switch lumx-switch--position-left lumx-switch--theme-light lumx-switch--is-unchecked"
>
  <div
    className="lumx-switch__input-wrapper"
  >
    <input
      aria-checked={false}
      className="lumx-switch__input-native"
      id="switch-uid"
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="lumx-switch__input-placeholder"
    >
      <div
        className="lumx-switch__input-background"
      />
      <div
        className="lumx-switch__input-indicator"
      />
    </div>
  </div>
</div>
`;
